src/util/file_log.rs

```diff
@@ -0,0 +1,225 @@
+        let ret = RotatingFileLogger {
+            level: logger::get_level_by_string(level),
+            core: Arc::new(RwLock::new(core)),
+        };
+        Ok(ret)
+    }
+
+}
+
+impl Log for RotatingFileLogger {
+    fn enabled(&self, metadata: &LogMetadata) -> bool {
+        metadata.level() <= self.level
+    }
+
+    fn log(&self, record: &LogRecord) {
+        if self.enabled(record.metadata()) {
+            let mut core = self.core.wl();
+            if core.should_rollover() {
+                core.do_rollover()
+            };
+            let now = time::now();
+            let _ = write!(core.file,
+                           "{},{:03} {}:{} - {:5} - {}\n",
+                           time::strftime("%Y-%m-%d %H:%M:%S", &now).unwrap(),
+                           now.tm_nsec / NANOSECONDS_PER_MILLISECOND,
+                           record.location().file().rsplit('/').nth(0).unwrap(),
+                           record.location().line(),
+                           record.level(),
+                           record.args());
+        }
+    }
+}
+
+impl Drop for RotatingFileLogger {
+    fn drop(&mut self) {
+        let mut core = self.core.wl();
+        core.close()
+    }
+}
+
+
+#[cfg(test)]
+mod tests {
+    extern crate log;
+    extern crate rand;
+    use time::{self, Timespec};
+    use std::io::prelude::*;
+    use std::fs::{OpenOptions, File};
+    use tempdir::TempDir;
+    use super::*;
+
+    extern crate utime;
+
+    fn check_log_file_with_single_line(file_path: &str, line: String) {
+        let mut f = File::open(file_path).unwrap();
+        let mut s = String::new();
+        let _ = f.read_to_string(&mut s);
+        let slice = &s;
+        if slice.len() < line.len() + 1 {
+            assert_eq!(slice, &line);
+        } else {
+            let read = &slice[slice.len() - line.len() - 1..slice.len() - 1];
+            assert_eq!(read, &line);
+        }
+    }
+
+    #[test]
+    fn test_rotating_file_logger() {
+        let level = "info";
+        let log_file = TempDir::new("test_rotating_file_logger.log")
+            .unwrap()
+            .path()
+            .to_str()
+            .unwrap()
+            .to_string();
+        // create a file with mtime == one day ago
+        {
+            let mut file = OpenOptions::new()
+                .append(true)
+                .create(true)
+                .open(&log_file)
+                .unwrap();
+            file.write_all(b"hello world!").unwrap();
+        }
+        let ts = time::now().to_timespec();
+        let one_day_ago = Timespec::new(ts.sec - ONE_DAY_SECONDS as i64, ts.nsec);
+        let rollover_time = compute_rollover_time(time::at(one_day_ago));
+        let time_in_sec = one_day_ago.sec as u64;
+        utime::set_file_times(&log_file, time_in_sec, time_in_sec).unwrap();
+        // init the log facility
+        init(level, &log_file).unwrap();
+        let line = "test log";
+        warn!("{}", line);
```

Comment:
all test cases run ok. The previous CI check falis on report like "address has been used", which seems unrelated.

@BusyJay Your idea of keeping test case of `RotatingFileLogger` independent of global log state make sence. But I couldnot find a way to contruct a `LogRecord` that has private fields. Could you take a look on it and give some help?